src/push.rs

```
@@ -0,0 +1,148 @@
+
+    for (ln, lv) in grouping {
+        // TODO: check label name
+        if lv.contains('/') {
+            return Err(Error::Msg(format!("value of grouping label {} contains '/': {}", ln, lv)));
+        }
+        url_components.push(ln.to_owned());
+        url_components.push(lv.to_owned());
+    }
+
+    push_url = format!("{}/metrics/job/{}", push_url, url_components.join("/"));
+
+    let encoder = TextEncoder::new();
+    let mut buf = Vec::new();
+    try!(encoder.encode(&mfs, &mut buf));
+
+    let request = HTTP_CLIENT.request(Method::from_str(method).unwrap(), &push_url)
+        .header(ContentType(encoder.format_type().parse().unwrap()))
+        .body(buf.as_slice());
+
+    let response = try!(request.send().map_err(|e| Error::Msg(format!("{}", e))));
+    if response.status == StatusCode::Accepted {
+        Ok(())
+    } else {
+        Err(Error::Msg(format!("unexpected status code {} while pushing to {}",
+                               response.status,
+                               push_url)))
+    }
+}
+
+/// `hostname_grouping_key` returns a label map with the only entry
+/// {instance="<hostname>"}. This can be conveniently used as the grouping
+/// parameter if metrics should be pushed with the hostname as label. The
+/// returned map is created upon each call so that the caller is free to add more
+/// labels to the map.
+pub fn hostname_grouping_key() -> HashMap<String, String> {
+    // Host names are limited to 255 bytes.
+    //   ref: http://pubs.opengroup.org/onlinepubs/7908799/xns/gethostname.html
+    let max_len = 256;
+    let mut name = vec![0u8; max_len];
+    match unsafe {
+        gethostname(name.as_mut_ptr() as *mut libc::c_char,
+                    max_len as libc::size_t)
+    } {
+        0 => {
+            let last_char = name.iter().position(|byte| *byte == 0).unwrap_or(max_len);
+            labels!{
+                "instance".to_owned() => str::from_utf8(&name[..last_char]).unwrap().to_owned(),
```

go client ignores the error, I think we should check it.